### PR TITLE
DynamicData array Length property

### PR DESCRIPTION
**Pull request by annelo-msft**
Addresses https://github.com/Azure/azure-sdk-for-net/issues/35944

- [ ] Add `Length` dynamic pseudo-property for JSON arrays

### CODE REVIEW

File: sdk/core/Azure.Core/tests/public/JsonDataArrayTests.cs [hunk 1]
```diff
@@ -234,14 +234,50 @@ public void CanSetObjectMemberViaArrayIndex()
         public void CanGetArrayLength()
         {
             dynamic data = JsonDataTestHelpers.CreateFromJson("[1, 2, 3]");
-            Assert.AreEqual(3, ((int[])data).Length);
+            Assert.AreEqual(3, data.Length);
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/tests/public/JsonDataArrayTests.cs:237]: We don't have to cast to an array type to access Length anymore.

File: sdk/core/Azure.Core/src/DynamicData/MutableJsonElement.cs [hunk 2]
```diff
@@ -91,6 +91,20 @@ public bool TryGetProperty(string name, out MutableJsonElement value)
...
 
+        public int GetArrayLength()
+        {
+            EnsureValid();
+
+            EnsureArray();
+
+            if (Changes.TryGetChange(_path, _highWaterMark, out MutableJsonChange change))
+            {
+                return change.AsJsonElement().GetArrayLength();
```
Comment (heaths) [inline sdk/core/Azure.Core/src/DynamicData/MutableJsonElement.cs:102]: Out of curiosity, does this get cached? I would expect that any array usage would probably require numerous bindings. Or is the lookup expected to be O(1)?

**Review by heaths — APPROVED** [review]

File: sdk/core/Azure.Core/tests/public/JsonDataArrayTests.cs [hunk 3]
```diff
@@ -234,14 +234,50 @@ public void CanSetObjectMemberViaArrayIndex()
...
+        }
+
+        [Test]
```
Comment (christothes) [inline sdk/core/Azure.Core/tests/public/JsonDataArrayTests.cs:240]: We should consider adding tests for getting Length when the data is not an array and also not an array but exposes a `Length` property.

**Review by christothes — APPROVED** [review]

**Review by JoshLove-msft — APPROVED** [review]
This is pretty cool - I assume we will document this via samples?

File: sdk/core/Azure.Core/src/DynamicData/MutableJsonElement.cs (hunk 2 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/src/DynamicData/MutableJsonElement.cs:102]: Great question, and I don't know offhand, but added to the list of perf follow-ups here: https://github.com/Azure/azure-sdk-for-net/issues/33856

File: sdk/core/Azure.Core/tests/public/JsonDataArrayTests.cs (hunk 3 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core/tests/public/JsonDataArrayTests.cs:240]: Good idea - will add.

**Pull request merged**